### ISSUE TITLE
New option to ignore preference and enforce only functional components.

### ISSUE THREAD

**Issue by cphoover** [yannickcr/eslint-plugin-react#2511]
With the advent of [hooks](https://reactjs.org/docs/hooks-intro.html) you no longe require class-based components to handle state. There should be an option to enforce using only function components and forbidding the use of class-based components altogether.

**Comment by ljharb** [comment]
Such a rule would also need a way to block functional components, since hooks interfere with testing and occasionally, readability.

I'm not sure this would be a valuable rule in either case.

**Comment by cphoover** [comment]
I disagree with hooks impacting readability (personally I think functional components are way more easy to reason about). However I agree If you would have an option for one convention (e.g. functional components only) there should also be a rule for the opposite (class components only).

**Comment by cphoover** [comment]
If there is an appetite for this change in `eslint-plugin-react` I can try and find some time to contribute this week.

**Comment by ljharb** [comment]
I think #2414 which fixes #690 might address this, which would make this a duplicate of #690?

**Comment by cphoover** [comment]
@ljharb unless this disallows class component it does not provide the desired feature.

**Comment by ljharb** [comment]
I can't conceive of how it'd be practical to have a rule that disallows class components (which are quite often more readable than hooks-based components, based on my experience) unless it also had an autofix that seamlessly converted in both directions between the two.

If you want to try to build that, I'd be happy to review it - but I suspect that the nature of JS as a language would make it pretty impossible.

**Comment by GeorgeTaveras1231** [comment]
> I can't conceive of how it'd be practical to have a rule that disallows class components (which are quite often more readable than hooks-based components, based on my experience) unless it also had an autofix that seamlessly converted in both directions between the two.
> If you want to try to build that, I'd be happy to review it - but I suspect that the nature of JS as a language would make it pretty impossible.

What makes this difficult is the desire to have autofix in place. I think having an optional rule that enforces the use of functional components is both useful and reasonably attainable. Autofix would be a nice to have but I wouldn't consider it a blocker. You can take one of two heuristical approaches to detect violation of this rule. 
1. Look for a class with a `#render` method that returns JSX. (Won't always work since its possible to delegate the return value to another function [Such as `render() { return this.renderThing() }`] , and recursively flattening functions to determine the final return value will be difficult).
2. Look for a `ClassDeclaration` with a `superClass` of `Component` (Referring to `babel-eslint9` ast demonstrated in astexplorer.net) -- and resolving the component reference to determine if it's imported from 'react'. This approach also has limitations (dynamic superclass? via `class MyComponent extends Mixin(React.Component)` -- this is not a common pattern anyways).

Nonetheless, I'm surprised by your pushback @ljharb. I understand you personally feel like class-based components are more readable, but honestly, this is the first I hear someone say that -- most people I've talked to about this and most articles I've read suggest the opposite is the popular opinion on the matter.

**Comment by robertjk** [comment]
> I can't conceive of how it'd be practical to have a rule that disallows class components (which are quite often more readable than hooks-based components, based on my experience) unless it also had an autofix that seamlessly converted in both directions between the two.

For example in my company we have a team of people who have been developing in React for no more than 2-3 years. Hardly any of us has ever used class components and understands them. I'd like to prevent having class components in our code base mostly for that reason - most of them team wouldn't be able to understand them.

**Comment by ljharb** [comment]
@robertjk if your team truly doesn't use or understand class components, when would they create one such that a rule would warn on them?

**Comment by robertjk** [comment]
> @robertjk if your team truly doesn't use or understand class components, when would they create one such that a rule would warn on them?

Likely they won't, but new developers join and I'd like to ensure no one in the future creates any class components in the code base.

We actually had one developer who was the opposite - he used only class components all his career, also he didn't understand function components and hooks. He was developing one project on his own and he used class components exlusively in that project. He's not working with us anymore, but another person like this can possibly join in the future.

**Comment by ljharb** [comment]
If that developer would ignore convention, wouldn’t they just ignore the eslint rule?

For surfacing those issues tho, sure, but at that point you could also block React.Component from being accessed or imported.

**Comment by robertjk** [comment]
> If that developer would ignore convention, wouldn’t they just ignore the eslint rule?

Could be, but could be not. Depends if he would ignore the convention because:
1. He knows but consciously breaks it
2. He simply doesn't know the convention

And for me one of the linter's role is aiding in problem no. 2. Linter configuration is also communicating project/company's coding standards, so that we don't have to write long style guides and have it memorized and checking it all the time in code reviews. That tool should ease that for us.

Also, even if he's in situation no. 1 and consciously breaks it, it's more likely to be caught in code review, if he's putting some `// eslint-disable` comments, than when he can submit the code without that. I for example ask every single time in code review, why someone is ignoring some ESLint rule, when I spot such a comment.

> you could also block React.Component from being accessed or imported

That's a nice idea. I might use it to enforce that, until a more straightforward rule is written.

Still I think it would be good to have a more straightforward rule to enforce what type of components you want to have in your project. From my personal experience (and reading other people comments in this thread confirms that), many people use one of the styles of defining components but not the other, and would benefit from such a rule.

**Comment by ljharb** [comment]
While I agree with the benefit the rule would offer you, I'm very hesitant to introduce a rule that blindly blocks an important feature of the language/React. We don't have one that blocks createReactClass components either, despite those being deprecated and obsolete.

**Comment by robertjk** [comment]
> I'm very hesitant to introduce a rule that blindly blocks an important feature of the language/React.
Why do you say "blindly"? There are people, who might want to block it quite consciously.

ESLint core rules allow you to block almost any single part of the JavaScript langue (in the most complex form using [`no-restricted-syntax`](https://eslint.org/docs/rules/no-restricted-syntax)) rule; even the `for` loop if that's your wish. The philosophy of their rules seems to be - you're in control, and you can enforce any coding style you'd want. Do you think that's a bad approach - that the more configurable the rules are the better?
